velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+operators by assigning each of them a shard of spill files.
+
+Given above considerations, we might just need to have a query config limit to
+cap a spill file size in practice. For unsorted spill, as we can continuously
+append to the same spill file so this helps to prevent a spill file from
+growing too big. For sorted spill, as each spill file can only store one sorted
+run of data so the actual spill file size will be the minimum of spillable data
+size and query config limit. As for now, the query config limit is a hard file
+size limit which can be tuned based on IO characteristics of the underlying
+storage system but we don’t expect it needs too much tuning in practice.
+
+Spill Target Size
+^^^^^^^^^^^^^^^^^
+The spill target size determines how much data to spill each time run spilling.
+If it is too small, then we will interrupt the operator execution very often
+for spilling. And it might also cause generating too many small spill files as
+well. If it is set to too big, then we might slow down the operator execution
+by spilling too much data to disk. We have a query config to set the spill
+target size as a factor of the query memory limit. We might need to tune this
+parameter a bit in practice to see its impact on performance.
+
+Data Storage
+------------
+The spilling just needs the underlying storage system to store a number of
+named streamed bytes. It may or may not need the namespace support. If the
+storage system supports namespace, we could store the spilled files from a
+query in one directory and delete them all by one directory deletion at the
+Spark Driver for `Sapphire <https://prestodb.io/blog/2021/10/26/Scaling-with-Presto-on-Spark>`_ or `Presto Coordinator <https://prestodb.io/docs/current/overview/concepts.html#coordinator>`_ for `Prestissmo <https://github.com/prestodb/presto/blob/master/presto-native-execution/README.md>`_ when a query
+finishes. The Velox is also able to delete each file individually if the
+storage system doesn’t support namespace hierarchy. When a system crashes, it
+is likely that some spill files are left on the storage system so we need some
+sort of garbage collection support. For storage systems that support TTL, we
+can leverage that feature to implement the spill file garbage collection. If
+not, we might need to build a lightweight GC service running out of band.
+
+==================
+Spilling Algorithm
+==================
+
+Hash Aggregation
+----------------
+
+    .. image:: images/spill-aggregation-spill.png
+       :width: 300
+       :align: left
+
+    .. image:: images/spill-aggregation-restore.png
+       :width: 300
+       :align: right
+
+The hash aggregation operator stores the intermediate aggregation state in a
+hash table with one table entry per each group by key. Each table entry is
+stored as a row in the row container of the hash table. When spilling gets
+triggered on the operator, its Spiller will scan all the rows in the row
```

Comment:
> When spilling gets triggered on the operator, its Spiller will scan all the

When spilling is triggered, the operator's Spiller object scans ... to select a set of partitions with the most data that collectively satisfy / meet the spill target.